File: benchmarks/compare/raptor/go/gtfs.go
package raptor

// MaxSafeInteger is JS Number.MAX_SAFE_INTEGER (2^53 - 1). Used as the
// "infinity" arrival sentinel and as the default Transfer.endTime.
const MaxSafeInteger int64 = 9007199254740991

// StopID e.g. "NRW".
type StopID = string

// Time in seconds since midnight (may exceed 24 hours).
type Time = int64

// Duration in seconds.
type Duration = int64

// TripID is a GTFS trip_id.
type TripID = string

// ServiceID is a GTFS service_id.
type ServiceID = string

// DateNumber is a date stored as a number, e.g. 20181225.
type DateNumber = int

// DayOfWeek: Sunday = 0 .. Saturday = 6 (matching JS Date.getDay).
type DayOfWeek = int

// RouteID is the route signature string.
type RouteID = string

// StopTime is a GTFS stop time.
type StopTime struct {
	Stop          StopID
	ArrivalTime   Time
	DepartureTime Time
	PickUp        bool
	DropOff       bool
}

// Trip is a GTFS trip.
type Trip struct {
	TripID    TripID
	StopTimes []StopTime
	ServiceID ServiceID
	Service   *Service
}

// Transfer is a leg with a duration instead of departure / arrival times.
type Transfer struct {
	Origin      StopID
	Destination StopID
	Duration    Duration
	StartTime   Time
	EndTime     Time
}

// TimetableLeg is a leg with concrete stop times.
type TimetableLeg struct {
	Origin      StopID
	Destination StopID
	StopTimes   []StopTime
	Trip        *Trip
}

// AnyLeg is either a *TimetableLeg or a *Transfer.
type AnyLeg interface {
	legOrigin() StopID
	legDestination() StopID
}

func (t *Transfer) legOrigin() StopID          { return t.Origin }
func (t *Transfer) legDestination() StopID     { return t.Destination }
func (l *TimetableLeg) legOrigin() StopID      { return l.Origin }
func (l *TimetableLeg) legDestination() StopID { return l.Destination }

// Journey is a collection of legs with computed departure / arrival times.
type Journey struct {
	Legs          []AnyLeg
	DepartureTime Time
	ArrivalTime   Time
}

File: benchmarks/compare/raptor/rust/src/gtfs.rs
//! GTFS data types (port of src/gtfs/GTFS.ts).

use std::rc::Rc;

use crate::service::Service;

/// StopID e.g. "NRW".
pub type StopId = String;

/// Time in seconds since midnight (may be greater than 24 hours).
pub type Time = i64;

/// Duration in seconds.
pub type Duration = i64;

/// `Number.MAX_SAFE_INTEGER` (2^53 - 1). The "infinity" arrival sentinel and the
/// default `Transfer.endTime`.
pub const MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;

/// Sunday = 0, Monday = 1 ... Saturday = 6 (JS `Date.getDay`).
pub type DayOfWeek = u32;

/// Date stored as a number, e.g. 20181225.
pub type DateNumber = i64;

/// GTFS stop time.
#[derive(Debug, Clone, PartialEq)]
pub struct StopTime {
    pub stop: StopId,
    pub arrival_time: Time,
    pub departure_time: Time,
    pub pick_up: bool,
    pub drop_off: bool,
}

/// GTFS trip.
#[derive(Debug, Clone)]
pub struct Trip {
    pub trip_id: String,
    pub stop_times: Vec<StopTime>,
    pub service_id: String,
    pub service: Rc<Service>,
}

/// Leg with a duration instead of departure and arrival time.
#[derive(Debug, Clone, PartialEq)]
pub struct Transfer {
    pub origin: StopId,
    pub destination: StopId,
    pub duration: Duration,
    pub start_time: Time,
    pub end_time: Time,
}
